### PR TITLE
ARROW-12392: [C++] Restore asynchronous streaming CSV reader

### PR DESCRIPTION
**Pull request by westonpace**
This restores ARROW-11887.  The only difference is that I kept the async path out of RecordBatchReader.  I now know the file formats do not use RecordBatchReader for the read so there was no need to add it in there yet.

It also now properly consumes (using RunInSerialExecutor) the reader via the nested parallel paths in the synchronous scanner.

### CODE REVIEW

**Comment by westonpace** [comment]
Keeping in draft until #10064 is resolved or else it'll just reintroduce the R deadlock issue.

File: cpp/src/arrow/util/async_generator.h
Comment (lidavidm) [inline cpp/src/arrow/util/async_generator.h:1506]: It would be nice to explain the purpose of this class and whether it's reentrant or not (well, the question is moot presumably because it's blocking). This essentially 1) lets you parameterize a generator on an executor (which is guaranteed to resolve to a SerialExecutor) and then 2) synchronously blocks on the generator via the SerialExecutor, effectively running all work on the main thread?

File: cpp/src/arrow/util/async_generator_test.cc
Comment (lidavidm) [inline cpp/src/arrow/util/async_generator_test.cc:1141]: Just to check my understanding, one call here (when IsSlow()) will synchronously sleep on the CPU executor, attempt to transfer to the I/O executor (which is a no-op as the future is already complete), then transfer back to the CPU executor (also a no-op)?

File: cpp/src/arrow/dataset/scanner.h
Comment (lidavidm) [inline cpp/src/arrow/dataset/scanner.h:154]: So to make sure, SafeExecute/SafeVisit are just a way to safely integrate async readers while we still have SyncScanner?

Comment (lidavidm) [reply cpp/src/arrow/dataset/scanner.h:154]: (Note #10062 removes the ScanTask bindings from Python, leaving just CGlib.)

File: cpp/src/arrow/dataset/file_csv.cc
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_csv.cc:191]: So `executor` here will be a SerialExecutor. Isn't that a behavior change? Previously, `StreamingReader::Make` would use the CPU thread pool. Now we're having it use the main thread.

File: cpp/src/arrow/util/async_generator_test.cc
Comment (westonpace) [reply cpp/src/arrow/util/async_generator_test.cc]: N/A, code removed.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h:154]: Correct.  These methods can go away when `SyncScanner` goes away.  Removing the `ScanTask` bindings won't be enough (although I'm happy to hear they are removed) because `SyncScanner::ToTable` and the synchronous dataset write both rely on scanning being safe to call (ala nested deadlock) from a CPU executor thread.

File: cpp/src/arrow/dataset/file_csv.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/file_csv.cc:191]: `SerialExecutor` assumes the current thread which may be from the CPU thread pool.  So, for example, `SyncScanner::ToTable` will continue to run `SafeExecute` on the CPU thread pool.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h:154]: Also, this will probably only be needed for CSV.  With parquet & IPC you created the async version in parallel with the sync verison.  With CSV there is just one (async) code path.

File: cpp/src/arrow/dataset/file_csv.cc
Comment (lidavidm) [reply cpp/src/arrow/dataset/file_csv.cc:191]: Ahh, I see now - before, while a cpu_executor was passed around in StreamingReader, it wasn't actually used. (MakeStreamingReader accepts an executor, but never uses it.) Now it is used, but it'll be a SerialExecutor, so we remain single-threaded as desired.

**Review by lidavidm — COMMENTED** [review]
Just one nit that I missed before (sorry).

File: cpp/src/arrow/dataset/file_base.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc]: Fixed.  It's frustrating that clang-format is smart enough to add these but not smart enough to remove them.

**Comment by westonpace** [comment]
Rebased and addressed remaining PR nit.  Perhaps best to wait until after the release has been finalized before merging this as it does affect the sync path.

**Comment by westonpace** [comment]
Rebase.  Appveyor failure appears unrelated.  Now that 4.0 is past the RC phase I'd appreciate merge or re-review.

**Review by lidavidm — APPROVED** [review]
Originally I wanted to wait for #10166 since I was worried there would be a lot of conflicts. However it looks like #10166 needs some more work + the conflicts are limited to scanner_internal.h so maybe we should just merge this. CC @bkietz as a heads up.

**Pull request closed without merging**
